### PR TITLE
[SPARK-12398] Smart truncation of DataFrame / Dataset toString

### PR DESCRIPTION
**Pull request by dilipbiswal**
When a DataFrame or Dataset has a long schema, we should intelligently truncate to avoid flooding the screen with unreadable information.
// Standard output
[a: int, b: int]

// Truncate many top level fields
[a: int, b, string ... 10 more fields]

// Truncate long inner structs
[a: struct<a: Int ... 10 more fields>]

### CODE REVIEW

**Comment by dilipbiswal** [comment]
@rxin Hi Reynold, can you please take a look and let me know your comments. Thanks !!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 1]
```diff
@@ -66,6 +66,9 @@ abstract class DataType extends AbstractDataType {
   /** Readable string representation for the type. */
   def simpleString: String = typeName
 
+  /** Readable string representation for the type with truncation */
+  def simpleString(maxNumberFields: Int): String = simpleString
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: this should be private to sql

**Comment by rxin** [comment]
cc @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala (hunk 1 above)
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: @rxin Will make the change. Thanks !!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala [hunk 2]
```diff
@@ -278,6 +278,23 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
     s"struct<${fieldTypes.mkString(",")}>"
   }
 
+  override def simpleString(maxNumberFields: Int): String = {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala]: `private[sql]`

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 3]
```diff
@@ -1155,4 +1155,54 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     val primitiveUDF = udf((i: Int) => i * 2)
     checkAnswer(df.select(primitiveUDF($"age")), Row(44) :: Row(null) :: Nil)
   }
+
+  test("SPARK-12398 truncated toString") {
+    val df1 = Seq((1: Long, "row1": String)).toDF("id", "name")
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Nit: you don't really need the type ascriptions here.  You can just do `1L` if you want it to be a long.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 4]
```diff
@@ -1155,4 +1155,54 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
...
+    assert(df4.toString === "[c1: bigint, c2: struct<_1: bigint,_2: string>]")
+
+    val df5 = Seq((1: Long, Tuple2(1: Long, "val": String), 20.0: Double)).toDF("c1", "c2", "c3")
+    assert(df5.toString === "[c1: bigint, c2: struct<_1: bigint,_2: string> ... 1 more field]")
+
+    val df6 =
+      Seq((1: Long,
+        Tuple2(1: Long, "val": String),
+        20.0: Double, 1: Integer)).toDF("c1", "c2", "c3", "c4")
+    assert(df6.toString === "[c1: bigint, c2: struct<_1: bigint,_2: string> ... 2 more fields]")
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Its a little odd that there are spaces after the top level commas but not the ones inside of struct.

Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: @marmbrus Hi Michael, i noticed that too while i was coding this up. Wasn't sure if i should change this. Now i am :-)

**Comment by marmbrus** [comment]
Thanks, merging to master.

**Pull request closed without merging**